### PR TITLE
[Routing] UrlHelper to get absolute URL for a path

### PR DESCRIPTION
**Pull request by vudaltsov**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes
| Fixed tickets | n/a
| License       | MIT
| Doc PR        | todo

I noticed that I need to generate absolute urls quite often. For example when normalizing uploads in API. I found Twig's `absolute_url()` really helpful, but obviously `Symfony\Bridge\Twig\Extension\HttpFoundationExtension` cannot be used as a normalizer's argument service.

In this PR I propose to extract `HttpFoundationExtension::generateAbsoluteUrl` and `HttpFoundationExtension::generateRelativePath` to separate interfaces which could be used on their own. Although this could be just a final class helper, I thought that we might leave a possibility for decoration here. That's why I created interfaces.

- [x] Split `HttpFoundationExtension` into two interfaces
- [x] Deprecate `HttpFoundationExtension::generateAbsoluteUrl` and `HttpFoundationExtension::generateRelativePath`
- [x] Add service definitions
- [x] Fix tests
- [ ] Add docs

### CODE REVIEW

**Comment by javiereguiluz** [comment]
@vudaltsov could you please show a small before/after example ... and maybe also briefly list the changes that existing apps may need to do. Thanks.

**Comment by vudaltsov** [comment]
@javiereguiluz , sure. 

```php
final class UserApiNormalizer implements NormalizerInterface
{
    private $absoluteUrlGenerator;

    public function __construct(AbsoluteUrlGeneratorInterface $absoluteUrlGenerator)
    {
        $this->absoluteUrlGenerator = $absoluteUrlGenerator;
    }

    public function normalize($user, $format = null, array $context = array())
    {
        return [
            // ...
            'avatar' => $this->absoluteUrlGenerator->generateAbsoluteUrl($user->avatar()->path()),
            // ...
        ];
    }
}
```

No changes needed for existing apps. Everything will continue to work as is. But two useful helping services will appear.

**Comment by ro0NL** [comment]
`AbsoluteUrlGeneratorInterface` looks like a routing concern.. probably caused by the "generate" term. Im a bit skeptical about introducing these interfaces under these names. Do we want a contrac even?

Maybe `RequestContext::toAbsolute/RelativeUrl()`?

**Comment by stof** [comment]
naming them UrlGenerator will indeed create confusion with the routing component

**Comment by vudaltsov** [comment]
Agree, I will consider different naming today.

@ro0NL , `RequestContext` is not aware of the `RequestStack`, so it's not gonna work. This helper uses them both to get maximum information to generate an absolute url.

As I explained, I created contracts because you might need to decorate these services once you have some additional context for generating an absolute url.

**Comment by ro0NL** [comment]
My idea was `RequestContext::toAbsoluteUrl(Request $request = null)`. From a controller pov you have the request at hand.

**Comment by vudaltsov** [comment]
@ro0NL , I understand the idea of passing `Request` as an argument. We'll be able to generate paths for any request then. 

The problem is that DX is worse since almost always I will have to inject two services instead of one (`+RequestStack`) into non request aware services. Also I don't think it is a responsibility of the `RequestContext` to do this — it's just keeping and providing request data. And I also doubt that I will ever need this in controllers. Mostly in other services, used by commands and message handlers as well (see normalizer example above).

**Comment by vudaltsov** [comment]
Refactored to `UrlHelper` without interfaces. Looks better? I don't really like the word `helper`, honestly...

**Comment by linaori** [comment]
Why is it final without an interface? Mocking this in userland will be a pain due to its dependencies and internal behavior. If it's something developers shouldn't use, I'd expect it to be `@internal`.

**Comment by ro0NL** [comment]
one could pass a request stack with mocked requests, and any requestcontext. Im not sure utilities need interfaces.

**Comment by linaori** [comment]
That creates a massive overhead of objects that you have to bring into the correct state in order to test your function, while their internal behavior is of no concern to your particular test if mocking would be sufficient.

**Comment by ro0NL** [comment]
what about putting the interface in symfony/contracts? It's a bit weird for HttpFoundation (a specific concept) to introduce a polymorphic matter.

File: src/Symfony/Component/HttpFoundation/UrlHelper.php [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpFoundation;
+
+use Symfony\Component\Routing\RequestContext;
+
+final class UrlHelper
```
Comment (ro0NL) [inline src/Symfony/Component/HttpFoundation/UrlHelper.php]: `CurrentRequestUri`?

File: src/Symfony/Component/HttpFoundation/UrlHelper.php [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    private $requestStack;
+    private $requestContext;
+
+    public function __construct(RequestStack $requestStack, RequestContext $requestContext = null)
+    {
+        $this->requestStack = $requestStack;
+        $this->requestContext = $requestContext;
+    }
+
+    public function getAbsoluteUrl(string $path): string
```
Comment (ro0NL) [inline src/Symfony/Component/HttpFoundation/UrlHelper.php]: with above naming a `toAbsolute()` would be sufficient

**Comment by fabpot** [comment]
I would not create new abstraction for that. Best would be to add that to the Request class instead.

**Comment by vudaltsov** [comment]
@fabpot, this cannot be done in `Request` because we need `RequestContext` as a fallback. The idea of these helper utilities is to provide absolute urls even outside the request scope (commands, for example).

File: src/Symfony/Component/HttpFoundation/UrlHelper.php (hunk 1 above)
Comment (vudaltsov) [reply src/Symfony/Component/HttpFoundation/UrlHelper.php]: `CurrentRequestUri` does not sound like a service

**Comment by vudaltsov** [comment]
After discussing with @fabpot , moved to Routing, since the helper depends on the `RequestContext` which is a routing class.

**Comment by vudaltsov** [comment]
Btw, what's the best way to handle deprecations in `HttpFoundationExtension`?
Add `@expectedDeprecation`?

**Comment by vudaltsov** [comment]
@fabpot , looked for places in Symfony where I could replace some code with a helper call (except the Twig `HttpFoundationExtension`). Two looked relevant:

1. https://github.com/symfony/symfony/blob/5fe3701f0541590193b964e86a97d0960c1ecc8b/src/Symfony/Component/Security/Http/HttpUtils.php#L65 . 
2. https://github.com/symfony/symfony/blob/5fe3701f0541590193b964e86a97d0960c1ecc8b/src/Symfony/Bundle/FrameworkBundle/Controller/RedirectController.php#L131

But it seems that `UrlHelper` will not really help there :)

**Comment by vudaltsov** [comment]
The latest Travis build: https://travis-ci.org/symfony/symfony/builds/516563856

File: src/Symfony/Component/Routing/UrlHelper.php [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+namespace Symfony\Component\Routing;
+
+use Symfony\Component\HttpFoundation\RequestStack;
+
+/**
+ * A helper service for manipulating URLs within and outside the request scope.
+ *
+ * @author Valentin Udaltsov <udaltsov.valentin@gmail.com>
+ */
+final class UrlHelper
```
Comment (greg0ire) [inline src/Symfony/Component/Routing/UrlHelper.php]: Alternative names: `UrlManipulator`, `UrlBuilder`

File: src/Symfony/Bridge/Twig/Extension/HttpFoundationExtension.php [hunk 4]
```diff
@@ -24,13 +25,27 @@
  */
 class HttpFoundationExtension extends AbstractExtension
 {
-    private $requestStack;
-    private $requestContext;
+    private $urlHelper;
 
-    public function __construct(RequestStack $requestStack, RequestContext $requestContext = null)
+    public function __construct($urlHelper, RequestContext $requestContext = null)
```
Comment (derrabus) [inline src/Symfony/Bridge/Twig/Extension/HttpFoundationExtension.php]: I think we should add a doc block documenting the correct type of `$urlHelper`. Also, I would remove `$requestContext` from the signature to show that the user is not supposed to pass it anymore. You can still fetch it with `func_get_arg()` for the BC layer.

File: src/Symfony/Bridge/Twig/Extension/HttpFoundationExtension.php [hunk 5]
```diff
@@ -24,13 +25,27 @@
...
-        $this->requestStack = $requestStack;
-        $this->requestContext = $requestContext;
+        if ($urlHelper instanceof UrlHelper) {
+            $this->urlHelper = $urlHelper;
+
+            return;
+        }
+
+        if (!$urlHelper instanceof RequestStack) {
+            throw new \InvalidArgumentException(sprintf('The first argument must be an instance of "%s" or an instance of "%s".', UrlHelper::class, RequestStack::class));
```
Comment (derrabus) [inline src/Symfony/Bridge/Twig/Extension/HttpFoundationExtension.php]: You could throw a `TypeError`. This would be consistent withe previous behavior.

**Review by derrabus — CHANGES_REQUESTED** [review]
This change makes TwigBridge incompatible with Routing < 4.3. The composer.json file should reflect that.

File: src/Symfony/Component/Routing/UrlHelper.php (hunk 3 above)
Comment (vudaltsov) [reply src/Symfony/Component/Routing/UrlHelper.php]: `UrlManipulator` is nice. Let's see what other reviewers think.
`UrlBuilder` is bad as a stateless service name.

**Comment by vudaltsov** [comment]
All comments addressed. 
Marked `HttpFoundationExtensionTest` with `@group legacy` since now all tests are in `UrlHelperTest` class.

**Review by derrabus — APPROVED** [review]

File: src/Symfony/Bridge/Twig/Extension/HttpFoundationExtension.php [hunk 6]
```diff
@@ -39,8 +62,8 @@ public function __construct(RequestStack $requestStack, RequestContext $requestC
     public function getFunctions()
     {
         return [
-            new TwigFunction('absolute_url', [$this, 'generateAbsoluteUrl']),
-            new TwigFunction('relative_path', [$this, 'generateRelativePath']),
+            new TwigFunction('absolute_url', [$this->urlHelper, 'getAbsoluteUrl']),
+            new TwigFunction('relative_path', [$this->urlHelper, 'getRelativePath']),
```
Comment (stof) [inline src/Symfony/Bridge/Twig/Extension/HttpFoundationExtension.php]: Please use a twig runtime, or a method on the extension, as that are the 2 ways which can compile to the most efficient code in Twig. Passing other callables like here requires initializing extensions to register all functions again, to have access to the callable. We don't do this in Symfony.

In this case, I suggest keeping the extensions method (and not deprecating them)

File: src/Symfony/Component/Routing/UrlHelper.php [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Routing;
```
Comment (stof) [inline src/Symfony/Component/Routing/UrlHelper.php]: It is weird to have this class in Routing, when it required dependency is from HttpFoundation and an optional one on Routing.

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/Compiler/ExtensionPass.php [hunk 8]
```diff
@@ -79,7 +79,7 @@ public function process(ContainerBuilder $container)
             $container->removeDefinition('twig.controller.preview_error');
         }
 
-        if ($container->has('request_stack')) {
+        if ($container->has('router.url_helper')) {
```
Comment (dmaicher) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/Compiler/ExtensionPass.php]: @vudaltsov as discussed I think this change can cause some problems for apps that will only update TwigBundle?

File: src/Symfony/Component/Routing/UrlHelper.php (hunk 7 above)
Comment (stof) [reply src/Symfony/Component/Routing/UrlHelper.php]: btw, this is a BC break for the HttpFoundationExtension, as it becomes unusable without the routing component.

**Review by stof — CHANGES_REQUESTED** [review]

**Comment by vudaltsov** [comment]
@stof , thank you! Moved it back to `HttpFoundation`. The service will work without the routing component, thanx to `"router.request_context" on-invalid="ignore"`

**Review by dmaicher — APPROVED** [review]

File: src/Symfony/Bridge/Twig/Extension/HttpFoundationExtension.php [hunk 9]
```diff
@@ -24,13 +25,35 @@
...
 {
-    private $requestStack;
-    private $requestContext;
+    private $urlHelper;
 
-    public function __construct(RequestStack $requestStack, RequestContext $requestContext = null)
+    /**
+     * @param UrlHelper $urlHelper
+     */
+    public function __construct($urlHelper)
```
Comment (HeahDude) [inline src/Symfony/Bridge/Twig/Extension/HttpFoundationExtension.php]: You should not inject your helper here, create a new Runtime instead.

**Review by HeahDude — CHANGES_REQUESTED** [review]

File: src/Symfony/Bridge/Twig/Extension/HttpFoundationExtension.php (hunk 9 above)
Comment (vudaltsov) [reply src/Symfony/Bridge/Twig/Extension/HttpFoundationExtension.php]: As discussed, the gain is too small, so I'll keep it the way it is.

**Review by HeahDude — APPROVED** [review]

File: src/Symfony/Bridge/Twig/composer.json [hunk 10]
```diff
@@ -30,7 +30,7 @@
         "symfony/http-kernel": "~3.4|~4.0",
         "symfony/mime": "~4.3",
         "symfony/polyfill-intl-icu": "~1.0",
-        "symfony/routing": "~3.4|~4.0",
+        "symfony/routing": "~4.3",
```
Comment (xabbuh) [inline src/Symfony/Bridge/Twig/composer.json]: This doesn't look needed anymore. Maybe the constraints for the HttpFoundation component need to be changed instead?

File: src/Symfony/Bridge/Twig/Extension/HttpFoundationExtension.php [hunk 11]
```diff
@@ -24,13 +25,35 @@
...
+            $this->urlHelper = $urlHelper;
+
+            return;
+        }
+
+        if (!$urlHelper instanceof RequestStack) {
+            throw new \TypeError(sprintf('The first argument must be an instance of "%s" or an instance of "%s".', UrlHelper::class, RequestStack::class));
+        }
+
+        @trigger_error(sprintf('Passing an instance of "%s" as the first constructor argument is deprecated since Symfony 4.3, pass an instance of "%s" instead.', RequestStack::class, UrlHelper::class), E_USER_DEPRECATED);
```
Comment (xabbuh) [inline src/Symfony/Bridge/Twig/Extension/HttpFoundationExtension.php]: The message should mention the class that contains the constructor (like `as the first constructor argument of "%s" is`).

File: src/Symfony/Bridge/Twig/Extension/HttpFoundationExtension.php [hunk 12]
```diff
@@ -24,13 +25,35 @@
...
+
+        if (2 === \func_num_args()) {
+            @trigger_error(sprintf('Passing the second argument to the constructor is deprecated since Symfony 4.3, pass an instance of "%s" as the first argument instead.', UrlHelper::class), E_USER_DEPRECATED);
```
Comment (xabbuh) [inline src/Symfony/Bridge/Twig/Extension/HttpFoundationExtension.php]: I am not sure that we really need to trigger another deprecation here.

**Review by xabbuh — CHANGES_REQUESTED** [review]
we need to document the deprecations in the `UPDATE-4.3.md` and `UPGRADE-5.0` files

File: UPGRADE-4.3.md [hunk 13]
```diff
@@ -145,6 +145,12 @@ Security
    }
    ```
 
+TwigBridge
+==========
+
+ * Passing a `RequestStack` instance as the first argument to the `HttpFoundationExtension` constructor
+   is deprecated, pass a `Symfony\Component\HttpFoundation\UrlHelper` instance instead.
```
Comment (xabbuh) [inline UPGRADE-4.3.md]: I would reword this a bit:

> The `$requestStack` and `$requestContext` arguments of the `HttpFoundationExtension` have been deprecated. Pass a `UrlHelper` instance as the only argument instead.

File: UPGRADE-5.0.md [hunk 14]
```diff
@@ -364,6 +364,13 @@ TwigBundle
  * The default value (`false`) of the `twig.strict_variables` configuration option has been changed to `%kernel.debug%`.
  * The `transchoice` tag and filter have been removed, use the `trans` ones instead with a `%count%` parameter.
  * Removed support for legacy templates directories `src/Resources/views/` and `src/Resources/<BundleName>/views/`, use `templates/` and `templates/bundles/<BundleName>/` instead.
+ 
+TwigBridge
+----------
+
+ * Passing a `RequestStack` instance as the first argument to the `HttpFoundationExtension` constructor
+   now throws a `\TypeError`, pass a `Symfony\Component\HttpFoundation\UrlHelper` instance instead.
+ * Removed the second constructor argument of `HttpFoundationExtension`.
```
Comment (xabbuh) [inline UPGRADE-5.0.md]: I would merge both items like this:

> The `$requestStack` and `$requestContext` arguments of the `HttpFoundationExtension` have been removed. Pass a `UrlHelper` instance as the only argument instead.

File: src/Symfony/Bridge/Twig/CHANGELOG.md [hunk 15]
```diff
@@ -6,6 +6,8 @@ CHANGELOG
 
  * added the `form_parent()` function that allows to reliably retrieve the parent form in Twig templates
  * added the `workflow_transition_blockers()` function
+ * passing a `RequestStack` instance as the first argument to the `HttpFoundationExtension` constructor
```
Comment (xabbuh) [inline src/Symfony/Bridge/Twig/CHANGELOG.md]: > The `$requestStack` and `$requestContext` arguments of the `HttpFoundationExtension` have been deprecated. Pass a `UrlHelper` instance as the only argument instead.

**Comment by vudaltsov** [comment]
@xabbuh , done!

**Review by xabbuh — APPROVED** [review]

File: src/Symfony/Bundle/TwigBundle/composer.json [hunk 16]
```diff
@@ -19,7 +19,7 @@
         "php": "^7.1.3",
         "symfony/config": "~4.2",
         "symfony/twig-bridge": "^4.2",
```
Comment (stof) [inline src/Symfony/Bundle/TwigBundle/composer.json]: this one needs to be bumped to 4.3

Comment (vudaltsov) [reply src/Symfony/Bundle/TwigBundle/composer.json]: done!

File: src/Symfony/Bridge/Twig/Extension/HttpFoundationExtension.php [hunk 17]
```diff
@@ -24,13 +25,35 @@
...
+
+        @trigger_error(sprintf('Passing a "%s" instance as the first argument to the "%s" constructor is deprecated since Symfony 4.3, pass a "%s" instance instead.', RequestStack::class, __CLASS__, UrlHelper::class), E_USER_DEPRECATED);
+
+        if (2 === \func_num_args()) {
+            $requestContext = \func_get_arg(1);
+            if (!$requestContext instanceof RequestContext) {
+                throw new \TypeError(sprintf('The second argument must be an instance of "%s".', RequestContext::class));
+            }
+        } else {
+            $requestContext = null;
```
Comment (fabpot) [inline src/Symfony/Bridge/Twig/Extension/HttpFoundationExtension.php]: Can you move this init outside of the `if` statement? Like before line 47?

File: UPGRADE-4.3.md [hunk 18]
```diff
@@ -145,6 +145,11 @@ Security
    }
    ```
 
+TwigBridge
+==========
+
+ * The `$requestStack` and `$requestContext` arguments of the `HttpFoundationExtension` have been deprecated. Pass a `Symfony\Component\HttpFoundation\UrlHelper` instance as the only argument instead.
```
Comment (fabpot) [inline UPGRADE-4.3.md]: Can you change this sentence (and the other ones in the changelogs) to read like `deprecated the ...`?

**Review by fabpot — APPROVED** [review]
minor changes to be done before merge

**Comment by vudaltsov** [comment]
Ready!

**Comment by fabpot** [comment]
Thank you @vudaltsov.

**Comment by OskarStark** [comment]
Very nice feature Valentin 💥
Would you mind creating a docs PR please? 🙏
